### PR TITLE
integration specs for views with Capybara and fix n+1 problems

### PR DESCRIPTION
**Pull request by Omwomo**
### Fixing n+1 problems

- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user. Attached below are screenshots of app logs before and after fixing n+1 problems, respectively.

Screenshot (1104)

Screenshot (1105)

### Integration tests  with Capybara for each view

- Write tests for User index page to check the following:
   - username of all other users.
   - profile picture for each user.
   - number of posts each user has written.
   - When I click on a user, I am redirected to that user's show page.
 
- Write tests for user show page to check the following:
   - profile picture.
   - user's username.
   - number of posts the user has written.
   - user's bio.
   - user's first 3 posts.
   - a button that lets me view all of a user's posts.
   - When I click a user's post, it redirects me to that post's show page.
   - When I click to see all posts, it redirects me to the user's post's index page.
  
Write tests for User post index page to check the following:
   - the user's profile picture.
   - the user's username.
   - the number of posts the user has written.
   - a post's title.
   - some of the post's body.
   - the first comments on a post.
   - how many comments a post has.
   - how many likes a post has.
   - section for pagination if there are more posts than fit on the view.
   - When I click on a post, it redirects me to that post's show page.
  
Write tests for Post show page to check the following:

   - the post's title.
   - who wrote the post.
   - how many comments it has.
   - how many likes it has.
   - the post body.
   - the username of each commentor.
   - the comment each commentor left.

### CODE REVIEW

File: spec/features/user_show_spec.rb
Comment (LionRouge1) [inline spec/features/user_show_spec.rb:1]: - [ ] Nice work testing all the functionality of your application :clap: However the test for the **User Show** is not passing. Kindly make sure all the test are passing and green.
Screenshot from 2023-11-30 12-17-27

**Review by LionRouge1 — CHANGES_REQUESTED** [review]
Hi **Team**, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Solves N+1 problem when fetching all posts and their comments :heavy_check_mark: 
- Use `Capybara` gem for the integration test :heavy_check_mark: 
- All the written test are passing and green :x:

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb
Comment (Omwomo) [reply spec/features/user_show_spec.rb:1]: hello @LionRouge1 On my side, all the tests inside the user_show_spec.rb file are passing without any failures. Here are the tests results.
Screenshot (1108)

Screenshot (1109)

Comment (LionRouge1) [reply spec/features/user_show_spec.rb:1]: Hi @Omwomo, kindly run `rspec /spec` or `bundle exec rspec spec/`. It seems the error is related to the `id` column that you added when creating the post

Comment (Omwomo) [reply spec/features/user_show_spec.rb:1]: Hello @LionRouge1 Here are the results of running ` bundle exec rspec spec/ ` command as you have suggested in the reply and all the teats are passing without changing anything on the code, but if I try removing the id column then the tests for the user show index fails.

Screenshot (1112)

Comment (LionRouge1) [reply spec/features/user_show_spec.rb:1]: Hey @Omwomo, I recently cleared the project from my machine and re-cloned your repository. Unfortunately, I'm still encountering the same error. In an attempt to troubleshoot, I investigated the issue further. 
As mentioned in my previous comment, the error seems to stem from the addition of the `id` column to both user and post creation, where a post with a conflicting value already exists in the test_database.

Assigning the user and post to an instance variable will resolve the issue. Kindly check this suggestion :point_down: 
```rb
RSpec.describe 'User management', type: :feature do
  before do
    @user = FactoryBot.create(:user, name: 'Example User')
    @post = FactoryBot.create(:post, title: 'Hello', author: @user)
  end

     .....

      # Test user's post title redirects to the post's show page
      first_post = @post

      post_path = user_post_path(@user, first_post)

      expect(page).to have_link(first_post.title, href: post_path)

      # Test clicking a user's post title redirects to the post's show page
      click_link(first_post.title)
      expect(page).to have_current_path(post_path)
    end
  end
end
```

**Review by joel-kalema — APPROVED** [review]
Hi @Omwomo 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

# Congratulations! 🎉 


## Highlights

The project is running successfully 👍 
PR is well descriptive 👍 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me **@joel-kalema** in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**